src/main/java/com/scalar/cassy/scheduler/CassyClient.java

```diff
@@ -0,0 +1,132 @@
+import com.scalar.cassy.rpc.BackupRequest;
+import com.scalar.cassy.rpc.BackupResponse;
+import com.scalar.cassy.rpc.CassyGrpc;
+import com.scalar.cassy.rpc.OperationStatus;
+import io.grpc.Channel;
+import io.grpc.ManagedChannelBuilder;
+import java.util.Arrays;
+import java.util.Optional;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+public class CassyClient {
+  private final CassyGrpc.CassyBlockingStub blockingStub;
+
+  public CassyClient() {
+    this(
+        ManagedChannelBuilder.forAddress(
+                System.getenv().getOrDefault("CASSY_SCHEDULER_HOST", "localhost"),
+                Integer.parseInt(System.getenv().getOrDefault("CASSY_SCHEDULER_PORT", "20051")))
+            .usePlaintext());
+  }
+
+  public CassyClient(ManagedChannelBuilder<?> channelBuilder) {
+    Channel channel = channelBuilder.build();
+    blockingStub = CassyGrpc.newBlockingStub(channel);
+  }
+
+  public void takeClusterSnapshot(String clusterId, int timeout) throws Exception {
+    BackupRequest backupRequest =
+        BackupRequest.newBuilder()
+            .setClusterId(clusterId)
+            .setBackupType(BackupType.CLUSTER_SNAPSHOT.get())
+            .build();
+    BackupResponse backupResponse = blockingStub.takeBackup(backupRequest);
+    startTask(backupResponse, timeout);
+  }
+
+  public void takeNodeSnapshot(String clusterId, int timeout, Optional<String[]> targetIps)
+      throws InterruptedException, TimeoutException, ExecutionException {
+    BackupRequest backupRequest =
+        BackupRequest.newBuilder()
+            .setClusterId(clusterId)
+            .setBackupType(BackupType.NODE_SNAPSHOT.get())
+            .build();
+    if (targetIps.isPresent()) {
+      backupRequest =
+          BackupRequest.newBuilder(backupRequest)
+              .addAllTargetIps(Arrays.stream(targetIps.get()).collect(Collectors.toList()))
+              .build();
+    }
+
+    startTask(blockingStub.takeBackup(backupRequest), timeout);
+  }
+
+  public void takeIncrementalBackup(String clusterId, int timeout, Optional<String[]> targetIps)
+      throws InterruptedException, TimeoutException, ExecutionException {
+    String snapshotId = "";
+    BackupListingResponse backupListingResponse =
+        blockingStub.listBackups(BackupListingRequest.newBuilder().setClusterId(clusterId).build());
+    for (BackupListingResponse.Entry entry : backupListingResponse.getEntriesList()) {
+      if ((entry.getBackupType() == BackupType.CLUSTER_SNAPSHOT.get()
+              || entry.getBackupType() == BackupType.NODE_SNAPSHOT.get())
+          && entry.getStatusValue() == OperationStatus.COMPLETED_VALUE) {
+        snapshotId = entry.getSnapshotId();
```

Comment:
I don't think it works in the following case.

Let's say it is a 3-node cluster (CLUSTER_ID: A), and
to level the load, snapshots are taken separately per node. 
Then, incremental backup is taken without `targetIps` specified (so taking for all the nodes).

|IP|SNAPSHOT_ID|STATUS|
|--|---|---|
|192.168.0.2|XXX|COMPLETED|
|192.168.0.3|YYY|COMPLETED|
|192.168.0.4|ZZZ|COMPLETED|

Reply:
@feeblefakie Sorry, it's been a while since I worked on this and I want to make sure I'm understanding your comment.

In the current implementation, when the target wants to take an Incremental Backup, we go through the following process:

1. Get a list of all the backups tied to a given `cluster_id`
2. Get the `snapshot_id` of the most recent snapshot in the list (cluster or node) that has a COMPLETED status value
3. Create an incremental backup starting from that snapshot
4. Send the backup request to the `target_ips`, or for all nodes if not specified

Even if the snapshots are taken separately per node, wouldn't this process still keep everything up to the latest changes since we are fetching the most recent snapshot and using it as a starting point for the backup?

Reply:
If they are taken separately, snapshot IDs are all different. so there is one previous snapshot ID for each IP.

Reply:
I see. So does this mean I should revise the code to perform backups for each `target_ip` in a cluster separately?

Reply:
@feeblefakie Could you explain a little more how you think I should handle this situation too?

Reply:
I think it should take a backup for each `target_ip` internally to avoid the above issue.
I said `internally` because the I/F of the tool might accept "take backups of all the IPs".

Reply:
Ok, thank you. What is `I/F` ?

Reply:
Sorry, it's interface.